tap_dynamodb/sync_strategies/log_based.py

```diff
@@ -100,65 +142,65 @@ def sync_log_based(config, state, stream):
-                        raise RuntimeError('Projection failed to apply: {}'.format(metadata.get(md_map, (), 'tap-mongodb.projection')))
-
-            record_message = singer.RecordMessage(stream=table_name,
-                                                  record=record_message,
-                                                  version=stream_version)
-            singer.write_message(record_message)
-
-            rows_saved += 1
-
-            seq_number_bookmarks[shard['ShardId']] = record['dynamodb']['SequenceNumber']
-            state = singer.write_bookmark(state, table_name, 'shard_seq_numbers', seq_number_bookmarks)
-
-            if rows_saved % WRITE_STATE_PERIOD == 0:
-                singer.write_state(state)
+        found_shards.append(shard['ShardId'])
+        # Only sync shards which we have not fully synced already
+        if shard['ShardId'] not in finished_shard_bookmarks:
+            rows_synced += sync_shard(shard, seq_number_bookmarks,
+                                      streams_client, stream_arn, projection, deserializer,
+                                      table_name, stream_version, state)
 
         # If the shard we just finished syncing is closed (i.e. has an
-        # EndingSequenceNumber), pop it off
+        # EndingSequenceNumber), remove it from the seq_number_bookmark
+        # and put it into the finished_shards bookmark
         if shard['SequenceNumberRange'].get('EndingSequenceNumber'):
             # Must check if the bookmark exists because if a shard has 0
             # records we will never set a bookmark for the shard
             if seq_number_bookmarks.get(shard['ShardId']):
                 seq_number_bookmarks.pop(shard['ShardId'])
+                finished_shard_bookmarks.append(shard['ShardId'])
+                state = singer.write_bookmark(state, table_name, 'finished_shards', finished_shard_bookmarks)
                 state = singer.write_bookmark(state, table_name, 'shard_seq_numbers', seq_number_bookmarks)
+                singer.write_state(state)
 
-        singer.write_state(state)
+    for shard in finished_shard_bookmarks:
+        if shard not in found_shards:
+            # Remove this shard because its no longer appearing when we query for get_shards
+            finished_shard_bookmarks.remove(shard)
+            state = singer.write_bookmark(state, table_name, 'finished_shards', finished_shard_bookmarks)
 
-    return rows_saved
+    singer.write_state(state)
+
+    return rows_synced
 
 def has_stream_aged_out(config, state, stream):
     # TODO Check if the beginning sequence number can move. If it cannot this > check is useless
     seq_number_bookmarks = singer.get_bookmark(state, stream['tap_stream_id'], 'shard_seq_numbers')
+
+    # TODO I think this is backwards? If we don't have any
+    # seq_number_bookmarks then where do we begin with logs?
```

Comment:
Has this TODO been resolved?

Reply:
No I missed fixing this and added comments explaining the logic behind this function.